felix/dissimilarity.py

```diff
@@ -0,0 +1,188 @@
+import pickle
+from gerrypy.data.load import *
+from gerrypy.analyze.viz import *
+import numpy as np
+import pandas as pd
+import itertools
+
+# Main Functions
+def k_most_dissimilar(plans, num_blocks, k):
+    """ Given a set of plans represented as a list of list of lists,
+    returns the k most dissimilar plans
+    Args:
+        plans: a set of maps
+        num_blocks: the number of blocks in the state
+        k: the number of most distant maps to return
+    """
+    maps = convert_plans(plans, num_blocks)
+    dist_matrix = compute_dist_matrix(maps)
+    selected_maps = furthest_maps(dist_matrix, maps, k)
+    return convert_back(selected_maps)
+
+
+# Helper Functions
+def convert_plans(plans, num_blocks):
+    """
+    Converts plans from a list of maps, each of which contain a list of districts, each of
+    which contains a list of blocks in that district into a 3D binary numpy tensor,
+    with the first dimension representing the number of maps, the second dimension representing
+    the number of districts, and the third dimension a binary array of length (num_blocks),
+    where a 1 in the ith entry indicates that the district contains block i.
+    Args:
+        plans: the list of maps to convert
+        num_blocks: the largest number of blocks in a single district. Used for the conversion
+        to binary arrays.
+    """
+    possible_districts = pd.Series(np.arange(0, num_blocks))
+    converted_plans = []
+    for i in range(len(plans)):
+        plan = plans[i]
+        converted_plan = []
+        for j in range(len(plan)):
+            district = plan[j]
+            converted_plan.append(possible_districts.isin(district))
+        converted_plans.append(converted_plan)
+    return np.array(converted_plans).astype(int)
+
+
+def compute_dist_matrix(maps):
+    """
+    Given a set of maps, return distance matrix
+    Args:
+      maps: (ndarray) a 3 dimensional tensor constructed using convert_plans
+    """
+    transposed_maps = np.transpose(maps, (0, 2,1))
+    dimension = maps.shape[0]
+    dist_matrix = np.zeros((dimension, dimension))
+    count = 0
+    i = 0
+    j = 0
+
+    for two_maps in itertools.product(maps, transposed_maps):
+        if i >= j:
+            dist = two_map_distance(two_maps[0], two_maps[1])
+            dist_matrix[i][j] = dist
+            dist_matrix[j][i] = dist
+        count += 1
+        if count % dimension == 0:
+            i += 1
+        j = count % dimension
+
+    return dist_matrix
+
+def two_map_distance(map_1, map_2):
+    """
+    Return the distance between map_1 and map_2
+    Args:
+        map_1: (ndarray) 2 Dimensional ndarray representing a map
+        map_2: (ndarray) 2 Dimensional ndarray representing a map
+    """
+    new_map = map_1 @ map_2
+    return 1000 * (1 / np.mean(np.max(new_map, 1)))
+
+
+def furthest_maps(dist_matrix, maps, num_maps):
```

Comment:
Implementation suggestion: Here is my version of this function, keeping the same intended greedy algorithmic functionality :
```suggestion
def farthest_maps(dist, maps, k):
    output_indices = []
    output_indices.extend(np.unravel_index(np.argmax(dist), dist.shape))

    for i in range(2, k):
        dist_to_out = dist[output_indices, :]
        dist_to_out[:, output_indices] = 0.0

        sum_dist_to_out = np.sum(dist_to_out, axis=0)
        output_indices.append(np.argmax(sum_dist_to_out))
    
    output_maps = maps[output_indices, ...]
    assert (len(np.unique(output_maps, axis=0)) == k), "Re-using previously found maps"
    return output_maps
```